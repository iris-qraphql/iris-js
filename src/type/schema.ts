import { uniqBy } from 'ramda';

import type { IrisError } from '../error';
import { inspect, instanceOf } from '../utils/legacy';
import type { ObjMap } from '../utils/ObjMap';
import type { Maybe } from '../utils/type-level';

import type { IrisNamedType, IrisResolverType, IrisType } from './definition';
import { getNamedType, isDataType, isResolverType } from './definition';
import type { GraphQLDirective } from './directives';
import { isDirective, specifiedDirectives } from './directives';

/**
 * Test if the given value is a GraphQL schema.
 */
export function isSchema(schema: unknown): schema is IrisSchema {
  return instanceOf(schema, IrisSchema);
}

export function assertSchema(schema: unknown): IrisSchema {
  if (!isSchema(schema)) {
    throw new Error(`Expected ${inspect(schema)} to be a GraphQL schema.`);
  }
  return schema;
}

export class IrisSchema {
  description: Maybe<string>;

  // Used as a cache for validateSchema().
  __validationErrors: Maybe<ReadonlyArray<IrisError>>;

  private _queryType: Maybe<IrisResolverType>;
  private _mutationType: Maybe<IrisResolverType>;
  private _subscriptionType: Maybe<IrisResolverType>;
  private _directives: ReadonlyArray<GraphQLDirective>;
  private _typeMap: TypeMap;

  constructor(config: Readonly<GraphQLSchemaConfig>) {
    this.__validationErrors = config.assumeValid === true ? [] : undefined;

    this.description = config.description;

    this._queryType = config.query;
    this._mutationType = config.mutation;
    this._subscriptionType = config.subscription;
    this._directives = uniqBy(
      (x) => x.name,
      [...(config.directives ?? []), ...specifiedDirectives],
    );

    // To preserve order of user-provided types, we add first to add them to
    // the set of "collected" types, so `collectReferencedTypes` ignore them.
    const allReferencedTypes: Set<IrisNamedType> = new Set(config.types);

    config.types?.forEach((type) => {
      allReferencedTypes.delete(type);
      collectReferencedTypes(type, allReferencedTypes);
    });

    [this._queryType, this._mutationType, this._subscriptionType].forEach(
      (type) =>
        type ? collectReferencedTypes(type, allReferencedTypes) : undefined,
    );

    for (const directive of this._directives) {
      // Directives are not validated until validateSchema() is called.
      if (isDirective(directive)) {
        for (const arg of directive.args) {
          collectReferencedTypes(arg.type, allReferencedTypes);
        }
      }
    }

    // Storing the resulting map for reference by the schema.
    this._typeMap = Object.create(null);

    for (const namedType of allReferencedTypes) {
      if (namedType == null) {
        continue;
      }

      const typeName = namedType.name;
      if (!typeName) {
        throw new Error(
          'One of the provided types for building the Schema is missing a name.',
        );
      }

      if (this._typeMap[typeName] !== undefined) {
        throw new Error(
          `Iris Schema must contain uniquely named types but contains multiple types named "${typeName}".`,
        );
      }
      this._typeMap[typeName] = namedType;
    }
  }

  get [Symbol.toStringTag]() {
    return 'IrisSchema';
  }

  getQueryType(): IrisResolverType | undefined {
    return this._queryType ?? undefined;
  }

  getMutationType(): IrisResolverType | undefined {
    return this._mutationType ?? undefined;
  }

  getSubscriptionType(): IrisResolverType | undefined {
    return this._subscriptionType ?? undefined;
  }

  getTypeMap = (): TypeMap => this._typeMap;

  getType(name: string): IrisNamedType | undefined {
    return this.getTypeMap()[name];
  }

  getDirectives(): ReadonlyArray<GraphQLDirective> {
    return this._directives;
  }

  getDirective(name: string): Maybe<GraphQLDirective> {
    return this.getDirectives().find((directive) => directive.name === name);
  }
}

type TypeMap = ObjMap<IrisNamedType>;

export type IrisSchemaValidationOptions = {
  assumeValid?: boolean;
  assumeValidSDL?: boolean;
};

export interface GraphQLSchemaConfig extends IrisSchemaValidationOptions {
  description?: Maybe<string>;
  query?: Maybe<IrisResolverType>;
  mutation?: Maybe<IrisResolverType>;
  subscription?: Maybe<IrisResolverType>;
  types?: Maybe<ReadonlyArray<IrisNamedType>>;
  directives?: Maybe<ReadonlyArray<GraphQLDirective>>;
}

function collectReferencedTypes(
  type: IrisType,
  typeSet: Set<IrisNamedType>,
): Set<IrisNamedType> {
  const namedType = getNamedType(type);

  if (!typeSet.has(namedType)) {
    typeSet.add(namedType);
    if (isResolverType(namedType)) {
      const variants = namedType.variants();
      if (namedType.isVariantType()) {
        const fields = Object.values(variants[0]?.fields ?? {});
        for (const field of fields) {
          collectReferencedTypes(field.type, typeSet);
          for (const arg of field.args) {
            collectReferencedTypes(arg.type, typeSet);
          }
        }
      } else {
        for (const variant of variants) {
          if (variant.type) {
            collectReferencedTypes(variant.type, typeSet);
          }
        }
      }
    } else if (isDataType(namedType)) {
      namedType
        .variants()
        .flatMap((x) => Object.values(x.fields ?? {}))
        .forEach((field) => collectReferencedTypes(field.type, typeSet));
    }
  }

  return typeSet;
}
