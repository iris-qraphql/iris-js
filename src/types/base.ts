export type Printable = {
  toString: () => string;
  toJSON: () => string;
  decode: <V>(v: unknown) => V;
};
