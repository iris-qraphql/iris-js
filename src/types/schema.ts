import type { DirectiveNode, ParseOptions, Source } from 'graphql';
import { prop, uniqBy } from 'ramda';

import { typeCheckASTValue } from '../validation/typeCheckASTValue';
import { validateSDL } from '../validation/validate';

import type { IrisError } from '../error';
import { irisNodeError } from '../error';
import { parse } from '../parsing';
import type { TypeMap } from '../utils/collectTypeMap';
import { collectTypeMap } from '../utils/collectTypeMap';
import { inspect } from '../utils/legacy';
import type { IrisMaybe, Maybe } from '../utils/type-level';
import { notNill } from '../utils/type-level';

import type {
  ArgumentDefinitionNode,
  DirectiveDefinitionNode,
  FieldDefinitionNode,
  NamedTypeNode,
  Role,
  TypeDefinitionNode,
  TypeNode,
  VariantDefinitionNode,
} from './ast';
import type {
  IrisArgument,
  IrisField,
  IrisType,
  IrisVariant,
} from './definition';
import { IrisScalars, IrisTypeDefinition, IrisTypeRef } from './definition';
import {
  GraphQLDeprecatedDirective,
  GraphQLDirective,
  specifiedDirectives,
} from './directives';
import { IrisKind } from './kinds';

class IrisSchema {
  description: Maybe<string>;

  readonly query?: IrisTypeDefinition<'resolver'>;
  readonly mutation?: IrisTypeDefinition<'resolver'>;
  readonly subscription?: IrisTypeDefinition<'resolver'>;
  readonly directives: ReadonlyArray<GraphQLDirective>;
  readonly typeMap: TypeMap;

  // Used as a cache for validateSchema().
  __validationErrors: Maybe<ReadonlyArray<IrisError>>;

  constructor({
    description,
    query,
    mutation,
    subscription,
    types = [],
    directives = [],
  }: Readonly<IrisSchemaConfig>) {
    this.description = description;
    this.query = query;
    this.mutation = mutation;
    this.subscription = subscription;
    this.directives = uniqBy(prop('name'), [
      ...directives,
      ...specifiedDirectives,
    ]);

    this.typeMap = collectTypeMap(
      [query, mutation, subscription, ...types].filter(notNill),
      this.directives,
    );
  }

  get [Symbol.toStringTag]() {
    return 'IrisSchema';
  }

  getType = (name: string): IrisMaybe<IrisTypeDefinition> => this.typeMap[name];

  getDirective = (name: string): Maybe<GraphQLDirective> =>
    this.directives.find((directive) => directive.name === name);
}

export interface IrisSchemaConfig {
  description?: Maybe<string>;
  query?: IrisTypeDefinition<'resolver'>;
  mutation?: IrisTypeDefinition<'resolver'>;
  subscription?: IrisTypeDefinition<'resolver'>;
  types?: ReadonlyArray<IrisTypeDefinition>;
  directives?: ReadonlyArray<GraphQLDirective>;
}

export function buildSchema(
  source: string | Source,
  options?: ParseOptions,
): IrisSchema {
  const documentAST = parse(source, { noLocation: options?.noLocation });
  const errors = validateSDL(documentAST);

  if (errors.length !== 0) {
    throw new Error(errors.map((error) => error.message).join('\n\n'));
  }

  const directiveDefs: Array<DirectiveDefinitionNode> = [];
  const typeMap: Record<string, IrisTypeDefinition> = {};

  function lookupType<R extends Role>(
    node: NamedTypeNode | VariantDefinitionNode<R>,
  ): IrisTypeDefinition<R> {
    const name = node.name.value;
    const type = IrisScalars[name] ?? typeMap[name];

    if (type === undefined) {
      throw new Error(`Unknown type: "${name}".`);
    }

    return type as IrisTypeDefinition<R>;
  }

  function getWrappedType(node: TypeNode): IrisType {
    if (node.kind === IrisKind.LIST_TYPE) {
      return new IrisTypeRef('LIST', getWrappedType(node.type));
    }
    if (node.kind === IrisKind.MAYBE_TYPE) {
      return new IrisTypeRef('MAYBE', getWrappedType(node.type));
    }
    return lookupType(node);
  }

  function buildDirective(node: DirectiveDefinitionNode): GraphQLDirective {
    return new GraphQLDirective({
      name: node.name.value,
      description: node.description?.value,
      locations: node.locations.map(({ value }) => value as any),
      isRepeatable: node.repeatable,
      args: node.arguments?.map(buildArgument),
      astNode: node,
    });
  }

  function buildArgument(astNode: ArgumentDefinitionNode): IrisArgument {
    const type: any = getWrappedType(astNode.type);
    const name = astNode.name.value;
    return {
      name,
      type,
      description: astNode.description?.value,
      defaultValue: typeCheckASTValue(astNode.defaultValue, type),
      deprecationReason: getDeprecationReason(astNode),
      astNode,
    };
  }

  const buildField = <R extends Role>(
    field: FieldDefinitionNode<R>,
  ): IrisField<R> =>
    ({
      name: field.name.value,
      type: getWrappedType(field.type),
      description: field.description?.value,
      deprecationReason: getDeprecationReason(field),
      astNode: field,
      args: field.arguments?.map(buildArgument),
      toJSON: () => field.name.value,
    } as IrisField<R>);

  const buildVariant = <R extends Role>(
    astNode: VariantDefinitionNode<R>,
  ): IrisVariant<R> => {
    const name = astNode.name.value;
    const description = astNode.description?.value;
    const deprecationReason = getDeprecationReason(astNode);

    if (!astNode.fields) {
      return {
        name,
        description,
        deprecationReason,
        astNode,
        type: lookupType(astNode),
      };
    }

    return {
      name,
      description,
      deprecationReason,
      fields: Object.fromEntries(
        (astNode.fields ?? []).map((field) => [
          field.name.value,
          buildField(field),
        ]),
      ),
      astNode,
    };
  };

  const buildType = <R extends Role>(
    astNode: TypeDefinitionNode<R>,
  ): IrisTypeDefinition<R> =>
    new IrisTypeDefinition({
      role: astNode.role,
      name: astNode.name.value,
      description: astNode.description?.value,
      variants: () => astNode.variants.map(buildVariant),
      astNode,
    });

  documentAST.definitions.forEach((def) => {
    const name = def.name.value;
    switch (def.kind) {
      case IrisKind.TYPE_DEFINITION:
        typeMap[name] = IrisScalars[name] ?? buildType(def);
        break;
      case IrisKind.DIRECTIVE_DEFINITION:
        directiveDefs.push(def);
        break;
    }
  });

  const assertRoot = (
    operation: string,
    type: IrisTypeDefinition,
  ): IrisTypeDefinition<'resolver'> => {
    if (type && !(type.role === 'resolver' && type.isVariantType)) {
      throw irisNodeError(
        `${operation} root type must be Object type${
          operation === 'Query' ? '' : ' if provided'
        }, it cannot be ${inspect(type)}.`,
        type.astNode,
      );
    }
    return type as IrisTypeDefinition<'resolver'>;
  };

  return new IrisSchema({
    description: undefined,
    types: Object.values(typeMap),
    query: assertRoot('query', typeMap.Query),
    mutation: assertRoot('mutation', typeMap.Mutation),
    subscription: assertRoot('subscription', typeMap.Subscription),
    directives: directiveDefs.map(buildDirective),
  });
}

function getDeprecationReason(node: {
  readonly directives?: ReadonlyArray<DirectiveNode>;
}): IrisMaybe<string> {
  const directiveNode = node.directives?.find(
    (directive) => directive.name.value === GraphQLDeprecatedDirective.name,
  );

  if (directiveNode === undefined) {
    return undefined;
  }

  const reason = directiveNode?.arguments?.find(
    (arg) => arg.name.value === 'reason',
  )?.value;

  return (typeCheckASTValue(reason, IrisScalars.String) as string) ?? '';
}

export type { IrisSchema };
