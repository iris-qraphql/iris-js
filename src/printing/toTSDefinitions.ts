import type { Maybe } from 'graphql/jsutils/Maybe';

import type { IrisSchema } from '../types/schema';
import type { ASTReducer } from '../utils/visitor';
import { visit } from '../utils/visitor';

import { block, join, wrap } from './utils';

const mapping: Record<string, Maybe<string>> = {
  String: 'string',
  Int: 'number',
  Float: 'number',
};

const printTypesASTReducer: ASTReducer<string> = {
  Name: { leave: (node) => node.value },
  Document: { leave: (node) => join(node.definitions, '\n\n') },
  NamedType: { leave: ({ name }) => mapping[name] ?? name },
  ListType: { leave: ({ type }) => '[' + type + ']' },
  MaybeType: { leave: ({ type }) => `Maybe<${type}>` },
  FieldDefinition: {
    leave: ({ description, name, type, directives }) =>
      wrap('', description, '\n') +
      name +
      ': ' +
      type +
      wrap(' ', join(directives, ' ')),
  },
  TypeDefinition: {
    leave: ({ description, name, directives, variants }) =>
      wrap('', description, '\n') +
      join(
        [
          'export type',
          name,
          join(directives, ' '),
          wrap('= ', join(variants, ' | ')),
        ],
        ' ',
      ),
  },
  VariantDefinition: {
    leave: ({ name, fields }) =>
      fields === undefined ? name : block([`__typename: '${name}'`, ...fields]),
  },
};

const printFunctionsASTReducer: ASTReducer<string> = {
  Name: { leave: (node) => node.value },
  Document: { leave: (node) => join(node.definitions, '\n\n') },
  NamedType: { leave: ({ name }) => `iris${name}` },
  ListType: { leave: ({ type }) => '[' + type + ']' },
  MaybeType: { leave: ({ type }) => `irisMaybe(${type})` },
  FieldDefinition: {
    leave: ({ description, name, type, directives }) =>
      wrap('', description, '\n') +
      name +
      ': ' +
      type +
      wrap(' ', join(directives, ' ')),
  },
  TypeDefinition: {
    leave: ({ description, name, variants }) =>
      wrap('', description, '\n') +
      join([
        `export const iris${name} = oneOf<${name}>([${join(
          variants,
          ',\n',
        )}])`,
      ]),
  },
  VariantDefinition: {
    leave: ({ name, fields }) =>
      fields
        ? `irisVariant('${name}',${block([join(fields, ',\n')])})`
        : `iris${name}`,
  },
};

const inlineUtils = `
import type { Maybe } from '../scripts/utils';
import {
  irisFloat,
  irisInt,
  irisMaybe,
  irisString,
  irisVariant,
  oneOf,
} from '../scripts/utils';
`;

export const toTSDefinitions = (ast: IrisSchema): string =>
  inlineUtils +
  '\n\n' +
  visit(ast.document, printTypesASTReducer) +
  '\n\n' +
  visit(ast.document, printFunctionsASTReducer);
