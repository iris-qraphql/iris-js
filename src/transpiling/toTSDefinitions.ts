
import type { TypeDefinitionNode } from '../types/ast';
import type { IrisSchema } from '../types/schema';

export const toTSDefinitions = (
  schema: IrisSchema,
): string => {

  // const transpileField =
  //   (resolvers: any) =>
  //   (
  //     { description, type, args }: IrisField<'resolver'>,
  //     name: string,
  //   ): GraphQLFieldConfig<any, any> => ({
  //     description,
  //     type: transpileTypeRef(type),
  //     resolve: resolvers[name],
  //     args: args ? Object.fromEntries(args.map(transpileArgument)) : undefined,
  //   });

  // const transpileTypeRef = (
  //   type: IrisTypeRef,
  //   isMaybe?: boolean,
  // ): GraphQLOutputType => {
  //   const withMaybe = (t: GraphQLOutputType) =>
  //     isMaybe ? t : new GraphQLNonNull(t);
  //   switch (type.kind) {
  //     case 'MAYBE':
  //       return transpileTypeRef(type.ofType, true);
  //     case 'LIST':
  //       return withMaybe(new GraphQLList(transpileTypeRef(type.ofType)));
  //     case 'NAMED':
  //       return withMaybe(lookup(type.ofType.name));
  //   }
  // };


  const transpileTypeDefinition = (
    type: TypeDefinitionNode,
  ): string => {
    const { name } = type;

    return `type ${name.value} = {}`;
  };


  return Object.values(schema.types).map(transpileTypeDefinition).join('\n'); 
};
