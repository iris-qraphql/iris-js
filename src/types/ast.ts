import type { Kind } from 'graphql';
import { Token } from 'graphql';

import type { Source } from '../parsing/source';
import type { Maybe } from '../utils/type-level';

import { GQLKind, IrisDirectiveLocation, IrisKind } from './kinds';

export { specifiedScalarTypes } from 'graphql';

export type ASTNode =
  | NameNode
  | DocumentNode
  | ArgumentNode
  | StringValueNode
  | NullValueNode
  | DirectiveNode
  | NamedTypeNode
  | ListTypeNode
  | FieldDefinitionNode
  | ArgumentDefinitionNode
  | TypeDefinitionNode
  | VariantDefinitionNode
  | MaybeTypeNode;

export type ValueNode = NullValueNode | StringValueNode;

/**
 * @internal
 */
export const QueryDocumentKeys: {
  [NodeT in ASTNode as NodeT['kind']]: ReadonlyArray<keyof NodeT>;
} = {
  Name: [],
  Document: ['definitions'],
  Argument: ['name', 'value'],
  StringValue: [],
  NullValue: [],
  Directive: ['name', 'arguments'],
  NamedType: ['name'],
  ListType: ['type'],
  MaybeType: ['type'],
  FieldDefinition: ['description', 'name', 'type', 'directives'],
  InputValueDefinition: [
    'description',
    'name',
    'type',
    'defaultValue',
    'directives',
  ],
  TypeDefinition: ['description', 'name', 'directives', 'variants'],
  VariantDefinition: ['description', 'name', 'fields', 'directives'],
};

const kindValues = new Set<string>(Object.keys(QueryDocumentKeys));
/**
 * @internal
 */
export function isNode(maybeNode: any): maybeNode is ASTNode {
  const maybeKind = maybeNode?.kind;
  return typeof maybeKind === 'string' && kindValues.has(maybeKind);
}

export class Location {
  /**
   * The character offset at which this Node begins.
   */
  readonly start: number;

  /**
   * The character offset at which this Node ends.
   */
  readonly end: number;

  /**
   * The Token at which this Node begins.
   */
  readonly startToken: Token;

  /**
   * The Token at which this Node ends.
   */
  readonly endToken: Token;

  /**
   * The Source document the AST represents.
   */
  readonly source: Source;

  constructor(startToken: Token, endToken: Token, source: Source) {
    this.start = startToken.start;
    this.end = endToken.end;
    this.startToken = startToken;
    this.endToken = endToken;
    this.source = source;
  }

  get [Symbol.toStringTag]() {
    return 'Location';
  }

  toJSON(): { start: number; end: number } {
    return { start: this.start, end: this.end };
  }
}

// VALUE
export type StringValueNode = {
  readonly kind: Kind.STRING;
  readonly loc?: Location;
  readonly value: string;
  readonly block?: boolean;
};

export type NullValueNode = {
  readonly kind: Kind.NULL;
  readonly loc?: Location;
};

export type ArgumentNode = {
  readonly kind: Kind.ARGUMENT;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly value: ValueNode;
};

export type DirectiveNode = {
  readonly kind: Kind.DIRECTIVE;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentNode>;
};

/** Name */

export type NameNode = {
  readonly kind: Kind.NAME;
  readonly loc?: Location;
  readonly value: string;
};

/** Document */

export type DocumentNode = {
  readonly kind: IrisKind.DOCUMENT;
  readonly loc?: Location;
  readonly definitions: ReadonlyArray<TypeDefinitionNode>;
};

export { Token };

/** Type Reference */

export type WrapperKind = 'LIST' | 'MAYBE' | 'NAMED';

export type TypeNode = NamedTypeNode | ListTypeNode | MaybeTypeNode;

export type NamedTypeNode = {
  readonly kind: IrisKind.NAMED_TYPE;
  readonly loc?: Location;
  readonly name: NameNode;
};

export type ListTypeNode = {
  readonly kind: IrisKind.LIST_TYPE;
  readonly loc?: Location;
  readonly type: TypeNode;
};

export type MaybeTypeNode = {
  readonly kind: IrisKind.MAYBE_TYPE;
  readonly loc?: Location;
  readonly type: NamedTypeNode | ListTypeNode;
};

export type ArgumentDefinitionNode = {
  readonly kind: IrisKind.ARGUMENT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly defaultValue?: ValueNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
};

export type FieldDefinitionNode = {
  readonly kind: IrisKind.FIELD_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
};

export type VariantDefinitionNode = {
  readonly kind: IrisKind.VARIANT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
  readonly isTypeVariantNode: boolean;
  readonly name: NameNode;
  readonly fields?: ReadonlyArray<FieldDefinitionNode>;
};

export type TypeDefinitionNode = {
  readonly kind: IrisKind.TYPE_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
  readonly variants: ReadonlyArray<VariantDefinitionNode>;
};

export interface DirectiveDefinitionNode {
  readonly kind: IrisKind.DIRECTIVE_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentDefinitionNode>;
  readonly repeatable: boolean;
  readonly locations: ReadonlyArray<NameNode>;
}

export const isTypeDefinitionNode = (
  node: ASTNode,
): node is TypeDefinitionNode => node.kind === IrisKind.TYPE_DEFINITION;

export const isTypeVariantNode = ({ name, variants }: TypeDefinitionNode) => {
  const typeName = name.value;
  const [variant] = variants;

  if (variants.length === 0) {
    return true;
  }

  return (
    variants.length === 1 &&
    variant.name.value === typeName &&
    variant.fields !== undefined
  );
};

export const getRefTypeName = (node: TypeNode): NameNode => {
  switch (node.kind) {
    case IrisKind.NAMED_TYPE:
      return node.name;
    case IrisKind.LIST_TYPE:
      return getRefTypeName(node.type);
    case IrisKind.MAYBE_TYPE:
      return getRefTypeName(node.type);
  }
};

export const getVariant = (
  { variants }: TypeDefinitionNode,
  name?: string,
): Maybe<VariantDefinitionNode> =>
  name ? variants.find((v) => v.name.value === name) : variants[0];

export const isRequiredArgument = (arg: ArgumentDefinitionNode) =>
  arg.type.kind !== IrisKind.MAYBE_TYPE && arg.defaultValue === undefined;

export const getDeprecationReason = (
  node: ASTNode & Pick<TypeDefinitionNode, 'directives'>,
): Maybe<string> => {
  const deprecated = node.directives?.find(
    (d) => d.name.value === 'deprecated',
  );

  if (!deprecated) {
    return undefined;
  }

  const reason = deprecated.arguments?.[0]?.value as Maybe<StringValueNode>;
  return reason?.value ?? '';
};

type IrisDirective = {
  name: string;
  description?: string;
  locations: ReadonlyArray<IrisDirectiveLocation>;
  args: ReadonlyArray<ArgumentDefinitionNode>;
  astNode?: DirectiveDefinitionNode;
};

export const specifiedDirectives: ReadonlyArray<IrisDirective> = Object.freeze([
  {
    name: 'deprecated',
    description: 'Marks an element of a GraphQL schema as no longer supported.',
    locations: [
      IrisDirectiveLocation.FIELD_DEFINITION,
      IrisDirectiveLocation.ARGUMENT_DEFINITION,
      IrisDirectiveLocation.VARIANT_DEFINITION,
    ],
    args: [
      {
        kind: IrisKind.ARGUMENT_DEFINITION,
        name: { value: 'reason', kind: GQLKind.NAME },
        description: {
          kind: GQLKind.STRING,
          value:
            'Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).',
        },
        defaultValue: { value: '', kind: GQLKind.STRING },
        type: {
          kind: IrisKind.NAMED_TYPE,
          name: { value: 'string', kind: GQLKind.NAME },
        },
      },
    ],
  },
]);
