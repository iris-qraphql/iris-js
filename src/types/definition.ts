import type { GraphQLScalarType } from 'graphql';
import { specifiedScalarTypes } from 'graphql';
import { pluck } from 'ramda';

import { irisError } from '../error';
import { didYouMean, inspect, suggestionList } from '../utils/legacy';
import type { ObjMap } from '../utils/ObjMap';
import { keyMap, mapValue } from '../utils/ObjMap';
import type { Maybe } from '../utils/type-level';

import type {
  ArgumentDefinitionNode,
  FieldDefinitionNode,
  Role,
  TypeDefinitionNode,
  VariantDefinitionNode,
  WrapperKind,
} from './ast';
import { isTypeVariantNode } from './ast';
import type { Printable } from './base';
import { IrisTypeRefImp } from './type-ref';

export const stdScalars = keyMap(specifiedScalarTypes, ({ name }) => name);

export const scalarNames = Object.keys(stdScalars);

export const isSpecifiedScalarType = (type: IrisTypeDefinition): boolean =>
  Boolean(stdScalars[type.name]);

export const irisTypeRef = <K extends WrapperKind, T extends Printable>(
  kind: K,
  ofType: T,
) => new IrisTypeRefImp(kind, ofType);

export type IrisTypeRef<R extends Role = Role> =
  | IrisTypeRefImp<'LIST', IrisTypeRef<R>>
  | IrisTypeRefImp<'MAYBE', IrisTypeRef<R>>
  | IrisTypeRefImp<'NAMED', IrisTypeDefinition<R>>;

export const liftType = <R extends Role>(t: IrisTypeDefinition<R>) =>
  irisTypeRef<'NAMED', IrisTypeDefinition<R>>('NAMED', t);

type IrisNode = {
  name: string;
  description?: Maybe<string>;
  deprecationReason?: Maybe<string>;
  toJSON?: () => unknown;
};

export type ThunkObjMap<T> = Thunk<ObjMap<T>>;
export type Thunk<T> = (() => T) | T;

const isThunk = <T>(thunk: Thunk<T>): thunk is () => T =>
  typeof thunk === 'function';

export const resolveThunk = <T>(thunk: Thunk<T>): T =>
  isThunk(thunk) ? thunk() : thunk;

export type IrisArgument = IrisNode & {
  type: IrisTypeRef<'data'>;
  defaultValue?: unknown;
  astNode?: ArgumentDefinitionNode;
};

export const isRequiredArgument = (arg: IrisArgument): boolean =>
  arg.type.kind !== 'MAYBE' && arg.defaultValue === undefined;

export type IrisField<R extends Role = Role> = IrisNode & {
  astNode?: FieldDefinitionNode<R>;
  type: R extends 'data' ? IrisTypeRef<'data'> : IrisTypeRef<R>;
  args?: R extends 'data' ? never : ReadonlyArray<IrisArgument>;
};

export type IrisVariant<R extends Role = Role> = IrisNode & {
  astNode?: VariantDefinitionNode<R>;
  toJSON?: () => string;
  fields?: ObjMap<IrisField<R>>;
  type?: IrisTypeDefinition<R>;
};

type IrisTypeConfig<R extends Role> = {
  role: R;
  name: string;
  description?: Maybe<string>;
  variants: Thunk<ReadonlyArray<IrisVariant<R>>>;
  astNode?: Maybe<TypeDefinitionNode<R>>;
  scalar?: R extends 'data' ? GraphQLScalarType<any, any> : undefined;
};

export class IrisTypeDefinition<R extends Role = Role> {
  name: string;
  description: Maybe<string>;
  astNode: Maybe<TypeDefinitionNode<R>>;
  role: R;
  isVariantType: boolean;

  #thunkVariants: () => ReadonlyArray<IrisVariant<R>>;
  #resolvedVariants?: ReadonlyArray<IrisVariant<R>>;
  #scalar?: GraphQLScalarType;

  constructor(config: Readonly<IrisTypeConfig<R>>) {
    this.name = config.name;
    this.description = config.description;
    this.astNode = config.astNode;
    this.role = config.role;
    this.#thunkVariants = () => resolveThunk(config.variants);
    this.#scalar = config.scalar;
    this.isVariantType = config.astNode
      ? isTypeVariantNode(config.astNode)
      : false;
  }

  get [Symbol.toStringTag]() {
    return 'IrisTypeDefinition';
  }

  get boxedScalar() {
    if (this.#scalar) {
      return this.#scalar;
    }

    const [variant] = this.variants();
    return stdScalars[variant?.name];
  }

  variants = (): ReadonlyArray<IrisVariant<R>> => {
    if (this.#resolvedVariants) {
      return this.#resolvedVariants;
    }
    this.#resolvedVariants = this.#thunkVariants();

    return this.#resolvedVariants;
  };

  variantBy = (name?: string): IrisVariant<R> =>
    lookupVariant(this.name, this.variants(), name);

  toString = (): string => this.name;

  toJSON = (): string => this.toString();
}

const lookupVariant = <V extends { name: string }>(
  typeName: string,
  variants: ReadonlyArray<V>,
  name?: string,
): V => {
  if (!name) {
    if (variants.length !== 1) {
      throw irisError(
        `Object ${inspect(
          name,
        )} must provide variant name for type "${typeName}"`,
      );
    }
    return variants[0];
  }

  const variant = variants.find((v) => v.name === name);

  if (!variant) {
    throw irisError(
      `Data "${typeName}" cannot represent value: ${inspect(name)}` +
        didYouMean(
          'the variant',
          suggestionList(name, pluck('name', variants)),
        ),
    );
  }

  return variant;
};

export const IrisScalars = mapValue(
  stdScalars,
  (scalar) =>
    new IrisTypeDefinition({
      role: 'data',
      name: scalar.name,
      description: scalar.description,
      variants: [{ name: scalar.name }],
      scalar,
    }),
);
