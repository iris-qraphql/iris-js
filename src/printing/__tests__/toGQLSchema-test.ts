
import { buildSchema } from '../../types/schema';

import { toTSDefinitions } from '../toTSDefinitions';

const matchGQLSnapshot = (src: string) =>
  expect(toTSDefinitions(buildSchema(src))).toMatchSnapshot();

describe('toGQLSchema', () => {
  it('hello world App', () => {
    matchGQLSnapshot(`
    data  Query = {
      hello: String
    }
  `);
  });

  it('hello world App', () => {
    matchGQLSnapshot(`
      data Lifespan
        = Immortal {}
        | Limited { max: Int? }

      data God = {
        name: String
        lifespan: Lifespan
      }

      data Deity
        = God
        | Titan { name: String }
    `);
  });
});
