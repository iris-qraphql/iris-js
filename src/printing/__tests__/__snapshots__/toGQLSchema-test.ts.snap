// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toGQLSchema hello world App 1`] = `
"
import type { Maybe } from '../scripts/utils';
import {
  irisFloat,
  irisInt,
  irisMaybe,
  irisString,
  irisVariant,
  oneOf,
} from '../scripts/utils';


export type Query = {
  __typename: 'Query'
  hello: string
}

export const irisQuery = oneOf<Query>([irisVariant('Query',{
  hello: irisString
})])"
`;

exports[`toGQLSchema hello world App 2`] = `
"
import type { Maybe } from '../scripts/utils';
import {
  irisFloat,
  irisInt,
  irisMaybe,
  irisString,
  irisVariant,
  oneOf,
} from '../scripts/utils';


export type Lifespan = {
  __typename: 'Immortal'
} | {
  __typename: 'Limited'
  max: Maybe<number>
}

export type God = {
  __typename: 'God'
  name: string
  lifespan: Lifespan
}

export type Deity = God | {
  __typename: 'Titan'
  name: string
}

export const irisLifespan = oneOf<Lifespan>([irisVariant('Immortal',{

}),
irisVariant('Limited',{
  max: irisMaybe(irisInt)
})])

export const irisGod = oneOf<God>([irisVariant('God',{
  name: irisString,
  lifespan: irisLifespan
})])

export const irisDeity = oneOf<Deity>([irisGod,
irisVariant('Titan',{
  name: irisString
})])"
`;
